File: mcp/admin/mcp-admin.py
#!/usr/bin/env python3
"""Unified MCP API key admin CLI.

This utility merges key issuance/rotation and active-key lookup into a single
entrypoint with subcommands:
- create: issue or rotate a key for a username
- get: read active key metadata (latest record per username)
- purge: remove stale rotated keys while keeping active records

Datastore characteristics:
- Local JSONL flat file for operational simplicity
- Append-only writes to preserve full key history
- Exact schema per row: username, created_at, api_key

Active-key convention:
- For a given username, the row with the most recent created_at is active.
"""

from __future__ import annotations

import argparse
import json
import re
import secrets
import sys
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

# Lightweight netID-style username validation.
USERNAME_PATTERN = re.compile(r"^[a-z][a-z0-9_-]{0,31}$")

# Keep datastore co-located with the admin CLI scripts.
SCRIPT_DIR = Path(__file__).resolve().parent
DEFAULT_DB_PATH = SCRIPT_DIR / "api_keys.jsonl"


@dataclass(frozen=True)
class ApiKeyRecord:
    """One immutable key event row in the datastore."""

    username: str
    created_at: str
    api_key: str

    def to_json_line(self) -> str:
        """Serialize to compact JSON for JSONL append operations."""

        payload = {
            "username": self.username,
            "created_at": self.created_at,
            "api_key": self.api_key,
        }
        return json.dumps(payload, separators=(",", ":"))


def build_parser() -> argparse.ArgumentParser:
    """Build CLI parser with command-specific subparsers."""

    parser = argparse.ArgumentParser(
        description="Admin utilities for MCP API key create/get operations."
    )
    subparsers = parser.add_subparsers(dest="command", required=True)

    create_parser = subparsers.add_parser(
        "create",
        help="Issue or rotate an API key for a username.",
    )
    create_parser.add_argument(
        "username",
        help="NetID username to issue an API key for (for example: npho).",
    )
    create_parser.add_argument(
        "-f",
        "--force",
        action="store_true",
        help="Issue a new key even if the username already has one (rotation).",
    )
    create_parser.add_argument(
        "--db-file",
        type=Path,
        default=DEFAULT_DB_PATH,
        help=f"Path to JSONL datastore (default: {DEFAULT_DB_PATH}).",
    )

    get_parser = subparsers.add_parser(
        "get",
        help="Get active key metadata (latest row) for one or all users.",
    )
    get_parser.add_argument(
        "username",
        nargs="?",
        help="Optional netID username to query. Omit to list all users.",
    )
    get_parser.add_argument(
        "--show-key",
        action="store_true",
        help="Include raw active API key in output (default is redacted).",
    )
    get_parser.add_argument(
        "--db-file",
        type=Path,
        default=DEFAULT_DB_PATH,
        help=f"Path to JSONL datastore (default: {DEFAULT_DB_PATH}).",
    )

    purge_parser = subparsers.add_parser(
        "purge",
        help="Purge stale rotated keys while keeping active keys.",
    )
    purge_parser.add_argument(
        "usernames",
        nargs="*",
        help="Optional netID usernames to purge. Omit to scan all users.",
    )
    purge_parser.add_argument(
        "--db-file",
        type=Path,
        default=DEFAULT_DB_PATH,
        help=f"Path to JSONL datastore (default: {DEFAULT_DB_PATH}).",
    )

    return parser


def resolve_db_path(db_path: Path) -> Path:
    """Resolve datastore path to an absolute normalized location."""

    return db_path.expanduser().resolve()


def validate_username(username: str) -> None:
    """Enforce a practical netID-like username policy."""

    if USERNAME_PATTERN.fullmatch(username):
        return

    raise ValueError(
        "Invalid username. Expected pattern: "
        "start with a lowercase letter, then lowercase letters/digits/_/- "
        "(max 32 chars)."
    )


def utc_now_iso8601() -> str:
    """Return current UTC timestamp in ISO 8601 with Z suffix."""

    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def parse_created_at(created_at: str) -> datetime:
    """Parse ISO 8601 timestamps, including trailing Z form."""

    return datetime.fromisoformat(created_at.replace("Z", "+00:00"))


def generate_api_key() -> str:
    """Generate cryptographically secure URL-safe key material."""

    return secrets.token_urlsafe(32)


def parse_record(raw_line: str, db_path: Path, line_number: int) -> ApiKeyRecord | None:
    """Parse and validate one datastore row.

    Invalid rows are skipped with warnings so one bad line does not block admin
    workflows for otherwise valid records.
    """

    try:
        item = json.loads(raw_line)
    except json.JSONDecodeError:
        print(
            f"Warning: ignored malformed JSON at {db_path}:{line_number}.",
            file=sys.stderr,
        )
        return None

    if not isinstance(item, dict):
        print(
            f"Warning: ignored non-object JSON at {db_path}:{line_number}.",
            file=sys.stderr,
        )
        return None

    required = {"username", "created_at", "api_key"}
    if set(item.keys()) != required:
        print(
            (
                "Warning: ignored row with unexpected schema at "
                f"{db_path}:{line_number}."
            ),
            file=sys.stderr,
        )
        return None

    try:
        username = str(item["username"])
        created_at = str(item["created_at"])
        api_key = str(item["api_key"])
        parse_created_at(created_at)
    except (TypeError, ValueError):
        print(
            (
                "Warning: ignored row with invalid field values at "
                f"{db_path}:{line_number}."
            ),
            file=sys.stderr,
        )
        return None

    return ApiKeyRecord(username=username, created_at=created_at, api_key=api_key)


def username_exists(db_path: Path, username: str) -> bool:
    """Return True if datastore already contains any row for username."""

    if not db_path.exists():
        return False

    with db_path.open("r", encoding="utf-8") as handle:
        for line_number, line in enumerate(handle, start=1):
            raw = line.strip()
            if not raw:
                continue

            record = parse_record(raw, db_path, line_number)
            if record is not None and record.username == username:
                return True

    return False


def append_record(db_path: Path, record: ApiKeyRecord) -> None:
    """Append one new record to JSONL datastore."""

    db_path.parent.mkdir(parents=True, exist_ok=True)

    with db_path.open("a", encoding="utf-8") as handle:
        handle.write(record.to_json_line())
        handle.write("\n")


def load_active_records(db_path: Path) -> dict[str, ApiKeyRecord]:
    """Load only the active row (latest created_at) per username."""

    active: dict[str, ApiKeyRecord] = {}
    if not db_path.exists():
        return active

    with db_path.open("r", encoding="utf-8") as handle:
        for line_number, line in enumerate(handle, start=1):
            raw = line.strip()
            if not raw:
                continue

            record = parse_record(raw, db_path, line_number)
            if record is None:
                continue

            current = active.get(record.username)
            if current is None or parse_created_at(record.created_at) > parse_created_at(
                current.created_at
            ):
                active[record.username] = record

    return active


def load_db_entries(db_path: Path) -> list[tuple[str, ApiKeyRecord | None]]:
    """Load datastore rows preserving raw line content and parsed records."""

    entries: list[tuple[str, ApiKeyRecord | None]] = []
    if not db_path.exists():
        return entries

    with db_path.open("r", encoding="utf-8") as handle:
        for line_number, line in enumerate(handle, start=1):
            raw_line = line.rstrip("\n")
            stripped = raw_line.strip()
            if not stripped:
                entries.append((raw_line, None))
                continue

            entries.append((raw_line, parse_record(stripped, db_path, line_number)))

    return entries


def format_get_output(record: ApiKeyRecord, show_key: bool) -> str:
    """Format stable JSON output for get command."""

    payload = {
        "username": record.username,
        "created_at": record.created_at,
        "api_key": record.api_key if show_key else "<redacted>",
    }
    return json.dumps(payload, separators=(",", ":"))


def run_create(args: argparse.Namespace) -> int:
    """Handle create subcommand lifecycle."""

    db_path = resolve_db_path(args.db_file)

    try:
        validate_username(args.username)
        exists = username_exists(db_path, args.username)
        if exists and not args.force:
            raise RuntimeError(
                "Username already exists in datastore. "
                "Use --force (or -f) to rotate and append a new key."
            )

        record = ApiKeyRecord(
            username=args.username,
            created_at=utc_now_iso8601(),
            api_key=generate_api_key(),
        )
        append_record(db_path, record)
    except (ValueError, RuntimeError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1
    except OSError as exc:
        print(f"Error: failed to write datastore: {exc}", file=sys.stderr)
        return 1

    print(f"username: {record.username}")
    print(f"created_at: {record.created_at}")
    print(f"api_key: {record.api_key}")
    print("status: rotated" if args.force else "status: issued")
    print(f"db_file: {db_path}")

    return 0


def run_get(args: argparse.Namespace) -> int:
    """Handle get subcommand lifecycle."""

    db_path = resolve_db_path(args.db_file)

    try:
        if args.username is not None:
            validate_username(args.username)
    except ValueError as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    if not db_path.exists():
        print(f"Error: datastore not found at {db_path}", file=sys.stderr)
        return 1

    active_records = load_active_records(db_path)

    if args.username is not None:
        record = active_records.get(args.username)
        if record is None:
            print(f"Error: no records found for username {args.username}", file=sys.stderr)
            return 1

        print(format_get_output(record, args.show_key))
        return 0

    for username in sorted(active_records):
        print(format_get_output(active_records[username], args.show_key))

    return 0


def run_purge(args: argparse.Namespace) -> int:
    """Handle purge subcommand lifecycle."""

    db_path = resolve_db_path(args.db_file)

    try:
        for username in args.usernames:
            validate_username(username)
    except ValueError as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    if not db_path.exists():
        print(f"Error: datastore not found at {db_path}", file=sys.stderr)
        return 1

    try:
        entries = load_db_entries(db_path)
    except OSError as exc:
        print(f"Error: failed to read datastore: {exc}", file=sys.stderr)
        return 1

    counts: dict[str, int] = {}
    for _, record in entries:
        if record is None:
            continue
        counts[record.username] = counts.get(record.username, 0) + 1

    if args.usernames:
        target_users = set(args.usernames)
    else:
        target_users = set(counts)

    rotated_users = sorted(
        username for username in target_users if counts.get(username, 0) > 1
    )
    total_removable = 0

    if rotated_users:
        print("Users with rotated keys eligible for purge:")
        for username in rotated_users:
            removable = counts[username] - 1
            total_removable += removable
            print(f"- {username}: {removable} rotated keys can be removed")
    else:
        print("No rotated keys found for purge scope.")

    print(f"Total rotated keys removable: {total_removable}")

    if total_removable == 0:
        print("Nothing to purge; exiting without modifying datastore.")
        return 0

    try:
        response = input("Proceed with purge? [Y/N]: ").strip().lower()
    except EOFError:
        response = "n"
    if response != "y":
        print("Purge cancelled.")
        return 0

    keep_index: dict[str, int] = {}
    keep_created_at: dict[str, datetime] = {}
    for index, (_, record) in enumerate(entries):
        if record is None:
            continue
        if record.username not in target_users:
            continue
        if counts.get(record.username, 0) <= 1:
            continue

        record_time = parse_created_at(record.created_at)
        current = keep_created_at.get(record.username)
        if current is None or record_time > current:
            keep_created_at[record.username] = record_time
            keep_index[record.username] = index

    filtered_lines: list[str] = []
    removed = 0
    for index, (raw_line, record) in enumerate(entries):
        if record is None:
            filtered_lines.append(raw_line)
            continue

        if (
            record.username in target_users
            and counts.get(record.username, 0) > 1
            and keep_index.get(record.username) != index
        ):
            removed += 1
            continue

        filtered_lines.append(raw_line)

    temp_path = db_path.with_suffix(f"{db_path.suffix}.tmp")
    try:
        with temp_path.open("w", encoding="utf-8") as handle:
            for line in filtered_lines:
                handle.write(line)
                handle.write("\n")
        temp_path.replace(db_path)
    except OSError as exc:
        print(f"Error: failed to write datastore: {exc}", file=sys.stderr)
        try:
            temp_path.unlink(missing_ok=True)
        except OSError:
            pass
        return 1

    print(f"Purged rotated keys: {removed}")
    print(f"db_file: {db_path}")
    return 0


def main() -> int:
    """CLI entrypoint."""

    parser = build_parser()
    args = parser.parse_args()

    if args.command == "create":
        return run_create(args)
    if args.command == "get":
        return run_get(args)
    if args.command == "purge":
        return run_purge(args)

    parser.print_help()
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: mcp/admin/mcp-admin.go
package main

import (
	"bufio"
	"bytes"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"time"
)

var usernamePattern = regexp.MustCompile(`^[a-z][a-z0-9_-]{0,31}$`)

type apiKeyRecord struct {
	Username  string `json:"username"`
	CreatedAt string `json:"created_at"`
	APIKey    string `json:"api_key"`
}

func main() {
	os.Exit(run(os.Args[1:], os.Stdout, os.Stderr))
}

func run(args []string, stdout io.Writer, stderr io.Writer) int {
	if len(args) == 0 {
		printRootUsage(stderr)
		return 1
	}

	command := args[0]
	if command == "-h" || command == "--help" || command == "help" {
		printRootUsage(stdout)
		return 0
	}

	switch command {
	case "create":
		config, err := parseCreateArgs(args[1:])
		if err != nil {
			if errors.Is(err, errHelpRequested) {
				printCreateUsage(stdout)
				return 0
			}
			fmt.Fprintf(stderr, "Error: %v\n", err)
			printCreateUsage(stderr)
			return 1
		}
		return runCreate(config, stdout, stderr)
	case "get":
		config, err := parseGetArgs(args[1:])
		if err != nil {
			if errors.Is(err, errHelpRequested) {
				printGetUsage(stdout)
				return 0
			}
			fmt.Fprintf(stderr, "Error: %v\n", err)
			printGetUsage(stderr)
			return 1
		}
		return runGet(config, stdout, stderr)
	default:
		fmt.Fprintf(stderr, "Error: unknown command %q\n", command)
		printRootUsage(stderr)
		return 1
	}

}

type createConfig struct {
	Username string
	Force    bool
	DBFile   string
}

type getConfig struct {
	Username string
	HasUser  bool
	ShowKey  bool
	DBFile   string
}

var errHelpRequested = errors.New("help requested")

func parseCreateArgs(args []string) (createConfig, error) {
	config := createConfig{DBFile: defaultDBPath()}

	for index := 0; index < len(args); index++ {
		arg := args[index]
		switch arg {
		case "-h", "--help":
			return createConfig{}, errHelpRequested
		case "-f", "--force":
			config.Force = true
		case "--db-file":
			index++
			if index >= len(args) {
				return createConfig{}, errors.New("missing value for --db-file")
			}
			config.DBFile = args[index]
		default:
			if strings.HasPrefix(arg, "--db-file=") {
				config.DBFile = strings.TrimPrefix(arg, "--db-file=")
				continue
			}
			if strings.HasPrefix(arg, "-") {
				return createConfig{}, fmt.Errorf("unknown flag %s", arg)
			}
			if config.Username != "" {
				return createConfig{}, errors.New("create expects exactly one username")
			}
			config.Username = arg
		}
	}

	if config.Username == "" {
		return createConfig{}, errors.New("missing username")
	}

	return config, nil
}

func parseGetArgs(args []string) (getConfig, error) {
	config := getConfig{DBFile: defaultDBPath()}

	for index := 0; index < len(args); index++ {
		arg := args[index]
		switch arg {
		case "-h", "--help":
			return getConfig{}, errHelpRequested
		case "--show-key":
			config.ShowKey = true
		case "--db-file":
			index++
			if index >= len(args) {
				return getConfig{}, errors.New("missing value for --db-file")
			}
			config.DBFile = args[index]
		default:
			if strings.HasPrefix(arg, "--db-file=") {
				config.DBFile = strings.TrimPrefix(arg, "--db-file=")
				continue
			}
			if strings.HasPrefix(arg, "-") {
				return getConfig{}, fmt.Errorf("unknown flag %s", arg)
			}
			if config.HasUser {
				return getConfig{}, errors.New("get accepts at most one username")
			}
			config.Username = arg
			config.HasUser = true
		}
	}

	return config, nil
}

func printRootUsage(out io.Writer) {
	fmt.Fprintln(out, "Usage: mcp-admin.go <command> [options]")
	fmt.Fprintln(out)
	fmt.Fprintln(out, "Admin utilities for MCP API key create/get operations.")
	fmt.Fprintln(out)
	fmt.Fprintln(out, "Commands:")
	fmt.Fprintln(out, "  create      Issue or rotate an API key for a username.")
	fmt.Fprintln(out, "  get         Get active key metadata (latest row) for one or all users.")
}

func printCreateUsage(out io.Writer) {
	fmt.Fprintln(out, "Usage: mcp-admin.go create [--db-file PATH] [--force|-f] <username>")
}

func printGetUsage(out io.Writer) {
	fmt.Fprintln(out, "Usage: mcp-admin.go get [--db-file PATH] [--show-key] [username]")
}

func defaultDBPath() string {
	if _, sourceFile, _, ok := runtime.Caller(0); ok {
		return filepath.Join(filepath.Dir(sourceFile), "api_keys.jsonl")
	}

	executablePath, err := os.Executable()
	if err == nil {
		return filepath.Join(filepath.Dir(executablePath), "api_keys.jsonl")
	}

	return "api_keys.jsonl"
}

func resolveDBPath(path string) (string, error) {
	if strings.HasPrefix(path, "~") {
		homeDir, err := os.UserHomeDir()
		if err != nil {
			return "", err
		}

		switch {
		case path == "~":
			path = homeDir
		case strings.HasPrefix(path, "~/"):
			path = filepath.Join(homeDir, path[2:])
		}
	}

	absolutePath, err := filepath.Abs(path)
	if err != nil {
		return "", err
	}

	return filepath.Clean(absolutePath), nil
}

func validateUsername(username string) error {
	if usernamePattern.MatchString(username) {
		return nil
	}

	return errors.New(
		"Invalid username. Expected pattern: start with a lowercase letter, then lowercase letters/digits/_/- (max 32 chars).",
	)
}

func utcNowISO8601() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000000Z")
}

func parseCreatedAt(createdAt string) (time.Time, error) {
	return time.Parse(time.RFC3339Nano, createdAt)
}

func generateAPIKey() (string, error) {
	randomBytes := make([]byte, 32)
	if _, err := rand.Read(randomBytes); err != nil {
		return "", err
	}

	return base64.RawURLEncoding.EncodeToString(randomBytes), nil
}

func parseRecord(rawLine string, dbPath string, lineNumber int, stderr io.Writer) (*apiKeyRecord, error) {
	var generic any
	if err := json.Unmarshal([]byte(rawLine), &generic); err != nil {
		fmt.Fprintf(stderr, "Warning: ignored malformed JSON at %s:%d.\n", dbPath, lineNumber)
		return nil, nil
	}

	objectValue, ok := generic.(map[string]any)
	if !ok {
		fmt.Fprintf(stderr, "Warning: ignored non-object JSON at %s:%d.\n", dbPath, lineNumber)
		return nil, nil
	}

	if len(objectValue) != 3 {
		fmt.Fprintf(stderr, "Warning: ignored row with unexpected schema at %s:%d.\n", dbPath, lineNumber)
		return nil, nil
	}

	usernameValue, hasUsername := objectValue["username"]
	createdAtValue, hasCreatedAt := objectValue["created_at"]
	apiKeyValue, hasAPIKey := objectValue["api_key"]
	if !hasUsername || !hasCreatedAt || !hasAPIKey {
		fmt.Fprintf(stderr, "Warning: ignored row with unexpected schema at %s:%d.\n", dbPath, lineNumber)
		return nil, nil
	}

	username, usernameOK := usernameValue.(string)
	createdAt, createdAtOK := createdAtValue.(string)
	apiKey, apiKeyOK := apiKeyValue.(string)
	if !usernameOK || !createdAtOK || !apiKeyOK {
		fmt.Fprintf(stderr, "Warning: ignored row with invalid field values at %s:%d.\n", dbPath, lineNumber)
		return nil, nil
	}

	if _, err := parseCreatedAt(createdAt); err != nil {
		fmt.Fprintf(stderr, "Warning: ignored row with invalid field values at %s:%d.\n", dbPath, lineNumber)
		return nil, nil
	}

	return &apiKeyRecord{Username: username, CreatedAt: createdAt, APIKey: apiKey}, nil
}

func usernameExists(dbPath string, username string, stderr io.Writer) (bool, error) {
	file, err := os.Open(dbPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return false, nil
		}
		return false, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for lineNumber := 1; scanner.Scan(); lineNumber++ {
		raw := strings.TrimSpace(scanner.Text())
		if raw == "" {
			continue
		}

		record, err := parseRecord(raw, dbPath, lineNumber, stderr)
		if err != nil {
			return false, err
		}
		if record != nil && record.Username == username {
			return true, nil
		}
	}

	if err := scanner.Err(); err != nil {
		return false, err
	}

	return false, nil
}

func appendRecord(dbPath string, record apiKeyRecord) error {
	if err := os.MkdirAll(filepath.Dir(dbPath), 0o755); err != nil {
		return err
	}

	file, err := os.OpenFile(dbPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		return err
	}
	defer file.Close()

	encoded, err := json.Marshal(record)
	if err != nil {
		return err
	}

	if _, err := file.Write(append(encoded, '\n')); err != nil {
		return err
	}

	return nil
}

func loadActiveRecords(dbPath string, stderr io.Writer) (map[string]apiKeyRecord, error) {
	active := make(map[string]apiKeyRecord)

	file, err := os.Open(dbPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return active, nil
		}
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for lineNumber := 1; scanner.Scan(); lineNumber++ {
		raw := strings.TrimSpace(scanner.Text())
		if raw == "" {
			continue
		}

		record, err := parseRecord(raw, dbPath, lineNumber, stderr)
		if err != nil {
			return nil, err
		}
		if record == nil {
			continue
		}

		current, exists := active[record.Username]
		if !exists {
			active[record.Username] = *record
			continue
		}

		recordTime, err := parseCreatedAt(record.CreatedAt)
		if err != nil {
			return nil, err
		}
		currentTime, err := parseCreatedAt(current.CreatedAt)
		if err != nil {
			return nil, err
		}

		if recordTime.After(currentTime) {
			active[record.Username] = *record
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return active, nil
}

func formatGetOutput(record apiKeyRecord, showKey bool) (string, error) {
	type outputRecord struct {
		Username  string `json:"username"`
		CreatedAt string `json:"created_at"`
		APIKey    string `json:"api_key"`
	}

	output := outputRecord{
		Username:  record.Username,
		CreatedAt: record.CreatedAt,
		APIKey:    "<redacted>",
	}
	if showKey {
		output.APIKey = record.APIKey
	}

	var buffer bytes.Buffer
	encoder := json.NewEncoder(&buffer)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(output); err != nil {
		return "", err
	}

	return strings.TrimSpace(buffer.String()), nil
}

func runCreate(config createConfig, stdout io.Writer, stderr io.Writer) int {
	dbPath, err := resolveDBPath(config.DBFile)
	if err != nil {
		fmt.Fprintf(stderr, "Error: failed to resolve datastore: %v\n", err)
		return 1
	}

	if err := validateUsername(config.Username); err != nil {
		fmt.Fprintf(stderr, "Error: %v\n", err)
		return 1
	}

	exists, err := usernameExists(dbPath, config.Username, stderr)
	if err != nil {
		fmt.Fprintf(stderr, "Error: failed to read datastore: %v\n", err)
		return 1
	}
	if exists && !config.Force {
		fmt.Fprintln(stderr, "Error: Username already exists in datastore. Use --force (or -f) to rotate and append a new key.")
		return 1
	}

	apiKey, err := generateAPIKey()
	if err != nil {
		fmt.Fprintf(stderr, "Error: failed to generate API key: %v\n", err)
		return 1
	}

	record := apiKeyRecord{
		Username:  config.Username,
		CreatedAt: utcNowISO8601(),
		APIKey:    apiKey,
	}

	if err := appendRecord(dbPath, record); err != nil {
		fmt.Fprintf(stderr, "Error: failed to write datastore: %v\n", err)
		return 1
	}

	fmt.Fprintf(stdout, "username: %s\n", record.Username)
	fmt.Fprintf(stdout, "created_at: %s\n", record.CreatedAt)
	fmt.Fprintf(stdout, "api_key: %s\n", record.APIKey)
	if config.Force {
		fmt.Fprintln(stdout, "status: rotated")
	} else {
		fmt.Fprintln(stdout, "status: issued")
	}
	fmt.Fprintf(stdout, "db_file: %s\n", dbPath)

	return 0
}

func runGet(config getConfig, stdout io.Writer, stderr io.Writer) int {
	dbPath, err := resolveDBPath(config.DBFile)
	if err != nil {
		fmt.Fprintf(stderr, "Error: failed to resolve datastore: %v\n", err)
		return 1
	}

	if config.HasUser {
		if err := validateUsername(config.Username); err != nil {
			fmt.Fprintf(stderr, "Error: %v\n", err)
			return 1
		}
	}

	if _, err := os.Stat(dbPath); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			fmt.Fprintf(stderr, "Error: datastore not found at %s\n", dbPath)
			return 1
		}
		fmt.Fprintf(stderr, "Error: failed to read datastore: %v\n", err)
		return 1
	}

	activeRecords, err := loadActiveRecords(dbPath, stderr)
	if err != nil {
		fmt.Fprintf(stderr, "Error: failed to read datastore: %v\n", err)
		return 1
	}

	if config.HasUser {
		record, exists := activeRecords[config.Username]
		if !exists {
			fmt.Fprintf(stderr, "Error: no records found for username %s\n", config.Username)
			return 1
		}

		formatted, err := formatGetOutput(record, config.ShowKey)
		if err != nil {
			fmt.Fprintf(stderr, "Error: failed to encode output: %v\n", err)
			return 1
		}
		fmt.Fprintln(stdout, formatted)
		return 0
	}

	usernames := make([]string, 0, len(activeRecords))
	for username := range activeRecords {
		usernames = append(usernames, username)
	}
	sort.Strings(usernames)

	for _, username := range usernames {
		formatted, err := formatGetOutput(activeRecords[username], config.ShowKey)
		if err != nil {
			fmt.Fprintf(stderr, "Error: failed to encode output: %v\n", err)
			return 1
		}
		fmt.Fprintln(stdout, formatted)
	}

	return 0
}
